src/extensions/compression/deflate.rs

```diff
@@ -0,0 +1,289 @@
+        // * The negotiation offer contains multiple extension parameters with
+        //   the same name.
+        // * The server doesn't support the offered configuration.
+        'outer: for (_, offer) in
+            extensions::parse_header(extensions).iter().filter(|(k, _)| k == self.name())
+        {
+            let mut config =
+                DeflateConfig { compression: self.compression, ..DeflateConfig::default() };
+            let mut agreed = Vec::new();
+            let mut seen_server_no_context_takeover = false;
+            let mut seen_client_no_context_takeover = false;
+            let mut seen_client_max_window_bits = false;
+            for param in offer {
+                match param.name() {
+                    SERVER_NO_CONTEXT_TAKEOVER => {
+                        // Invalid offer with multiple params with same name is declined.
+                        if seen_server_no_context_takeover {
+                            continue 'outer;
+                        }
+                        seen_server_no_context_takeover = true;
+                        config.server_no_context_takeover = true;
+                        agreed.push(Param::new(SERVER_NO_CONTEXT_TAKEOVER));
+                    }
+
+                    CLIENT_NO_CONTEXT_TAKEOVER => {
+                        // Invalid offer with multiple params with same name is declined.
+                        if seen_client_no_context_takeover {
+                            continue 'outer;
+                        }
+                        seen_client_no_context_takeover = true;
+                        config.client_no_context_takeover = true;
+                        agreed.push(Param::new(CLIENT_NO_CONTEXT_TAKEOVER));
+                    }
+
+                    // Max window bits are not supported at the moment.
+                    SERVER_MAX_WINDOW_BITS => {
+                        // A server declines an extension negotiation offer with this parameter
+                        // if the server doesn't support it.
+                        continue 'outer;
+                    }
+                    // Not supported, but server may ignore and accept the offer.
+                    CLIENT_MAX_WINDOW_BITS => {
+                        // Invalid offer with multiple params with same name is declined.
+                        if seen_client_max_window_bits {
+                            continue 'outer;
+                        }
+                        seen_client_max_window_bits = true;
+                    }
+
+                    // Offer with unknown parameter MUST be declined.
+                    _ => {
+                        continue 'outer;
+                    }
+                }
+            }
+
+            return Some((to_header_value(&agreed), DeflateContext::new(Role::Server, config)));
+        }
+
+        None
+    }
+
+    pub(crate) fn accept_response(&self, agreed: &[Param]) -> Result<DeflateContext, DeflateError> {
```

Comment:
Oh, looks to something similar which I wrote in the comment above about `negotiation_response` :slightly_smiling_face: